FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/tls_freertos.c

```diff
@@ -0,0 +1,604 @@
+
+            if( mbedtlsError != 0 )
+            {
+                LogError( ( "Failed to set server name, with mbedTLS high level error %s"
+                            " and with low level error %s.",
+                            mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                            mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+
+                returnStatus = TLS_TRANSPORT_API_ERROR;
+            }
+        }
+    }
+
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        /* Perform the TLS handshake. */
+        do
+        {
+            mbedtlsError = mbedtls_ssl_handshake( &( pNetworkContext->sslContext.context ) );
+        } while( ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_READ ) ||
+                 ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_WRITE ) );
+
+        if( mbedtlsError != 0 )
+        {
+            LogError( ( "Failed to perform TLS handshake, with mbedTLS high level error %s"
+                        " and with low level error %s.",
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+
+            returnStatus = TLS_TRANSPORT_HANDSHAKE_FAILED;
+        }
+    }
+
+    if( returnStatus != TLS_TRANSPORT_SUCCESS )
+    {
+        sslContextFree( &( pNetworkContext->sslContext ) );
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) TLS handshake successful.",
+                   pNetworkContext ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static TlsTransportStatus_t initMbedtls( void )
+{
+    TlsTransportStatus_t returnStatus = TLS_TRANSPORT_SUCCESS;
+    int mbedtlsError = 0;
+
+    /* Set the mutex functions for mbed TLS thread safety. */
+    mbedtls_threading_set_alt( mbedtls_platform_mutex_init,
+                               mbedtls_platform_mutex_free,
+                               mbedtls_platform_mutex_lock,
+                               mbedtls_platform_mutex_unlock );
+
+    /* Initialize contexts for random number generation. */
+    mbedtls_entropy_init( &entropyContext );
+    mbedtls_ctr_drbg_init( &ctrDrgbContext );
+
+    /* Add a strong entropy source. At least one is required. */
+    mbedtlsError = mbedtls_entropy_add_source( &entropyContext,
+                                               mbedtls_platform_entropy_poll,
+                                               NULL,
+                                               32,
+                                               MBEDTLS_ENTROPY_SOURCE_STRONG );
+
+    if( mbedtlsError != 0 )
+    {
+        LogError( ( "Failed to add entropy source, with mbedTLS high level error %s"
+                    " and with low level error %s.",
+                    mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                    mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+        returnStatus = TLS_TRANSPORT_API_ERROR;
```

Comment:
This enum name is misleading (to me). I initially thought that it referred to an error cause of invalid API call from the caller...
Should it be named as `TLS_TRANSPORT_INTERNAL_ERROR` instead?

Reply:
Agree. Will update